libc/src/search/hsearch.cpp

```diff
@@ -0,0 +1,35 @@
+//===-- Implementation of hsearch -------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/search/hsearch.h"
+#include "src/__support/HashTable/table.h"
+#include "src/errno/libc_errno.h"
+#include "src/search/hsearch/global.h"
+
+namespace LIBC_NAMESPACE {
+LLVM_LIBC_FUNCTION(ENTRY *, hsearch, (ENTRY item, ACTION action)) {
+  ENTRY *result;
+
+  switch (action) {
+  case FIND:
+    result = internal::global_hash_table->find(item.key);
+    if (result == nullptr) {
+      libc_errno = ESRCH;
+    }
+    break;
+  case ENTER:
+    result = internal::global_hash_table->insert(item);
```

Comment:
what happens if someone calls `hsearch` before `hcreate` has been called, or calls `hsearch` after `hcreate` then `hdestroy` have been called?

Reply:
`POSIX.1` leaves it as unspecified. Per `MUSL`'s implementation, such cases (including memory leak caused by multiple hcreate) are not checked.

Reply:
@nickdesaulniers what do you think of adding sth like `LIBC_ASSERT(internal::global_hash_table != nullptr);`?

Reply:
Yeah, that's better than nothing IMO. Please add.

Reply:
FWIW, I was studying bionic last night and found their implementations of these come from FreeBSD.

https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/upstream-freebsd/lib/libc/stdlib/hcreate.c

Looks like they use a `bool` to track whether the global hash table hash been initialized.  They also do the right thing if search has been called before the table was ever explicitly created.  I think we should, too.  Do you mind following up with such a change?

Reply:
@nickdesaulniers sure. Looking into the code, they are actually doing resizing and merely regard the maximum number as a hint. In that way, if a global hashtable is not initialized, they will just create a minimal table and resize it during insertion. 

It requires some modifications but will not be too painful to add such features. I will be forming a patch by next Monday.